_posts/2019-05-06-live-view-with-presence.md

```diff
@@ -0,0 +1,587 @@
+  Presence.track(
+    self(),
+    topic(chat.id),
+    current_user.id,
+    %{
+      first_name: current_user.first_name,
+      email: current_user.email,
+      user_id: current_user.id
+    }
+  )
+  ...
+end
+```
+
+Here, we use the [`Presence.track/4`](https://hexdocs.pm/phoenix/Phoenix.Presence.html#c:track/4) function to track our live view process as a presence. We add the PID of the LiveView process to Presence's data store, along with a payload describing the new user under a topic of `"chat:#{chat.id}"` and a key of the user's ID.
+
+The Presence process's state for the given topic will look something like this:
+
+```elixir
+%{
+  "1" => %{
+    metas: [
+      %{
+        email: "sophie@email.com",
+        first_name: "Sophie",
+        phx_ref: "TNV4PzRfyhw=",
+        user_id: 1
+      }
+  }
+}
+```
+
+#### Broadcasting Presence To Existing Users
+
+When we call `Presence.track`, Presence will broadcast a `"presence_diff"` event over its PubSub backend. We told our Presence module to use the same PubSub server as the rest of the application––the very same server that backs our `PhatWeb.Endpoint`.
+
+Recall that our live view clients are subscribing to this PubSub server via the following call in the `mount/2` function: ` PhatWeb.Endpoint.subscribe(topic(chat.id))`. So, these subscribing LiveView processes will receive the `"presence_diff"` event, which looks something like this:
+
+```elixir
+%{
+  event: "presence_diff",
+  payload: %{
+    joins:
+      %{
+        "1" => %{
+          metas: [
+            %{
+              email: "sophie@email.com",
+              first_name: "Sophie",
+              phx_ref: "TNV4PzRfyhw=",
+              user_id: 1
+            }
+          }
+        },
+    leaves: %{},
+  }
+}
+```
+
+The event's payload will describe the users that are joining the channel when `Presence.track/4` is called. Although we will respond to the `"presence_diff"` event, we won't do anything with the event's payload for now. However, you could imagine using it to create custom user experiences such as welcoming the newly joined user or alerting existing users that a certain new member has joined the chat room.
+
+In order to respond to the event we'll define a `handle_info/2` function in our live view that will match the `"presence_diff"` event:
+
+```elixir
+# lib/phat_web/live/chat_live_view.ex
+
+def handle_info(%{event: "presence_diff"}, socket = %{assigns: %{chat: chat}}) do
+
+end
+```
+
+This function has two responsibilities:
+
+* Get the list of present users for the given chat room topic from the Presence data store
+* Update the LiveView socket's state to reflect this list of users
+
+```elixir
+def handle_info(%{event: "presence_diff", payload: _payload}, socket = %{assigns: %{chat: chat}}) do
+  users =   
+    Presence.list(topic(chat.id))
+    |> Enum.map(fn {_user_id, data} ->
+      data[:metas]
+      |> List.first()
+    end)
+
+  {:noreply,
```

Comment:
as a personal preference, I would turn this tuple to one-liner, but that's totally up to you.